### DISCUSSION TITLE
Fetching Data

### DISCUSSION THREAD

**Discussion by t-monaco** [next-safe-action/next-safe-action#282] [Q&A]
Hi community, I know that my be silly but nowadays with all the multiple posibilites of data fetching, this library is still valid/useful for data? 

I know that is recommended for mutations and validations, but to fetch data?

**Comment by TheEdoRan** [comment]
Wouldn't recommend to use it for fetching data. I'd wait for the more (supposedly) powerful React Server Functions to see if they'll support fetching data via GET method.

**Comment by X-Titouan** [comment]
Hey !

Any updates on this ?

We want to use next-safe-action as a wrapper to fetch our data (so we can reuse all the middlewares we created for permissions checks, etc) Normalize data output, logging etc.

However, when we try to Promise.all multiple get actions inside a server component, Next.js doesn't seem to parallelize them properly...

Any idea why, or how we could handle this better? How do you (@TheEdoRan ) do to handle this problem ?

Thanks a lot!

**Reply by TheEdoRan** [reply]
Hi, Server Actions aren't meant for that purpose. There are several drawbacks to using them that way, one of them, as you found out, being that they can't be parallelized, which is a React limitation, not a next-safe-action one. I highly suggest using Server Actions just for mutations, as the Next.js team recommends.

As for what I do, I fetch data in RSCs and/or using tRPC.

**Reply by enisze** [reply]
hey @TheEdoRan in which regard can't they be parallelized? e.g. using `Promise.all` or in which regard? :)
